### PR TITLE
Build docker & publish images

### PR DESCRIPTION
**Pull request by douglascamata**
# Why?

To allow users to easily run `obsctl` in an interactive environment through Docker without having to install Go or compile the project.

# What was done

* The Docker images are built using the same mechanism as in observatorium/up.
  * The `latest` tag is never used, only versioned ones.
* Fixed the `make build` command to use `go build` with the local source instead of `go install`.
* The project is now built by default with Go 1.18.

# Has to be done by maintainers

* Create Docker image repo at quay.io/observatorium/obsctl
* Add the Quay secrets to the Github Actions, like done for Up.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -42,7 +42,8 @@ all: format build
 .PHONY: build
 build: check-git deps ## Build obsctl.
 	@echo ">> building obsctl"
-	@GOBIN=$(GOBIN) go install github.com/observatorium/obsctl
+	@GOBIN=$(GOBIN) go build -v -ldflags '-w -extldflags '-static'' -o obsctl
```
Comment (saswatamcode) [inline Makefile:45]: Is there any existing issue that this addresses? 🙂 
I believe this makes it harder to test obsctl out, with having to manually add this executable to your path each time, you want to install the latest build.

We use `go install` in other CLIs too for eg in [mdox](https://github.com/bwplotka/mdox/blob/main/Makefile#L45), [bingo](https://github.com/bwplotka/bingo/blob/main/Makefile#L46) so I would prefer to keep this as a separate `make install` command if needed.

File: .github/workflows/publish.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      - "main"
+    tags:
+      - "v*"
+jobs:
+  image:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Login to image registry
+        uses: docker/login-action@v1
```
Comment (saswatamcode) [inline .github/workflows/publish.yaml:15]: How about we have commands like `make docker`,  `make docker-login`, and `make docker-push`. That way docker images can be built locally with ease and we can have simple commands in CI as well similar to how we do it in [observatorium/api](https://github.com/observatorium/api/blob/main/Makefile#L140).

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+WORKDIR /workspace
+
+COPY . .
+
+RUN make build
+
+FROM alpine:3.16.0
+
+COPY --from=builder /workspace/obsctl /usr/bin/obsctl
+COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
```
Comment (saswatamcode) [inline Dockerfile:14]: Maybe adding an `ENTRYPOINT` would be cool here?

File: .github/workflows/publish.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+          password: ${{ secrets.QUAY_PASSWORD }}
+      - name: Generate image names
+        id: images
+        run: |
+          echo -n ::set-output name=IMAGES::
+          tag=($echo "quay.io/observatorium/obsctl:$(git rev-parse --abbrev-ref HEAD | tr / -)-$(date +%Y-%m-%d)-$(git rev-parse --short HEAD)")
+          TAGS=($echo $(echo $tag | awk '{ print tolower($1) }'))
+          ( IFS=$','; echo "${TAGS[*]}" )
+      - name: Build and publish image(s) to registry
+        uses: docker/build-push-action@v2
```
Comment (saswatamcode) [inline .github/workflows/publish.yaml:28]: I wonder if there is a need for a multi-arch building step here. Especially as obsctl is meant to run on people's machines.

**Review by saswatamcode — COMMENTED** [review]
Thanks! Some comments.
Would also be cool if you can add a section to docs on how to use docker image e.g `Alternative Ways to Run`. 🙂

File: Makefile (hunk 1 above)
Comment (douglascamata) [reply Makefile:45]: `go install` has the side effect of replacing the binary in `$GOBIN`, which might not be what one expect when running `make build`. Building something is very different from installing, that's why we make the classic combination of `make build; make install`.

I'm happy move the `go install` to `make install`.

File: Dockerfile (hunk 3 above)
Comment (douglascamata) [reply Dockerfile:14]: `ENTRYPOINT` doesn't work well with a stateful command like `obsctl`, @saswatamcode. For the image to be useful, the user has to run a bunch of commands to setup the API, login with a tenant, change the context, and then finally use the metrics commands, for example.

File: .github/workflows/publish.yaml (hunk 4 above)
Comment (douglascamata) [reply .github/workflows/publish.yaml:28]: People's machines might be an Apple Silicon Macbook (arm64)... it's getting more and more common. 😄

**Comment by saswatamcode** [comment]
Hmm, so after some deeper thought, I'm confused what the utility of doing this is. 😅 

The utility of obsctl lies in the fact that once someone saves their API details and logs in as a tenant, they do not have to re-authenticate again, ever. They can now do as many operations (`obsctl metrics *`) as needed, without having to reconfigure their credentials or worry about refreshing some token. And they can save multiple APIs and tenants, and switch between them.

By containerizing obsctl, it loses the ability to persist this data, as containers are ephemeral by nature. So one would have to add API, login as a tenant, and then do the operations they need to do __every time__.  The obsctl config file is lost as soon as the operation completes. They also cannot do operations for different tenants at once.

This makes obsctl no different from `token-refresher` + `cURL` , so might as well use that instead. 😅 

To retain the usefulness of obsctl, a user would have to run obsctl containers with a volume mount flag to save the config file somewhere and use it every time they run an obsctl container, which feels more complicated and way less intuitive than just installing Go and building obsctl.

@douglascamata you mentioned that using a docker image instead of installing obsctl, made it easier to use. Would love to know what issues you found that made you think so? 🤔 
 
Feel free to correct me if I'm wrong somewhere! 🙂

**Comment by douglascamata** [comment]
If going through the setup process repeatedly is a problem, we can use the environment variable for custom config folder and mount it into the container. It's an easy solution that is already possible and requires only documentation.

File: Makefile (hunk 1 above)
Comment (moadz) [reply Makefile:45]: +1, if I'm building a binary to test it makes sense to have it replace my path for the executable

File: .github/workflows/publish.yaml (hunk 4 above)
Comment (moadz) [reply .github/workflows/publish.yaml:28]: +1 multiarch builds pls. Most of the team don't use M1 macs... :{

**Comment by moadz** [comment]
Just weighing in as I believe we're at an impasse. 

What utility does docker provide? Portable environment agnostic executables for a binary. Given this is being built using go, and we can do multi-arch builds, +obsctl sessions are stateful (that's kinda the whole point of obsctl) it doesn't make sense to wrap this in docker (no real benefits in portability and we add an additional step to our build processes) as we _lose_ utility (persistent OIDC config).

I'm not convinced this has to be dockerized, we should instead look at distributing the cross-platform binaries for easier consumption.

**Review by moadz — CHANGES_REQUESTED** [review]
This doesn't really adds any utility to obsctl that it doesn't currently have with go cross-platform binaries.

File: Makefile (hunk 1 above)
Comment (douglascamata) [reply Makefile:45]: @moadz do you mean it makes sense that a `make build` command replaces things in your path, forever? This has such a different meaning from `build`, this sounds like `install`.

**Comment by douglascamata** [comment]
The utility is described in the **why** section of the PR: `To allow users to **easily run obsctl** in an interactive environment through Docker **without having to install Go or compile the project**.`

> (...) that it doesn't currently have with go cross-platform binaries.

Currently this project does not distribute any binary.

**Comment by moadz** [comment]
> Currently this project does not distribute any binary.

I don't think using docker images as a vehicle for distributing a CLI is appropriate. There are other alternatives we can look into (Github Release Assets integrates with GH actions and is a CDN for cross-platform go binaries that users can WGET/curl to get on their local path) 

Docker is really really cubersome for this kind of tool. It has a prerequisite of a user needing to install docker, and it's easier to just give them a binary to run that is persistent between sessions.

**Pull request closed without merging**